### DISCUSSION TITLE
어디서부터 구현을 해야되나요?

### DISCUSSION THREAD

**Discussion by mirusu400** [dlunch/wie#285] [Q&A] [answered]
wipi 에뮬레이터를 찾다가 이 프로젝트를 찾았습니다.

저도 이 프로젝트에 참가를 해보고싶은데.. 코드를 봐도 어디서부터 시작해야될지 모르겠네요

KTF, LGT, SKT 따로 구현을 해야되기도하고 C, JAVA 구현체가 또 따로.. 게다가 JAVA는 직접 만드신 rust기반 jvm 위에서 돌아가는 것 같고요

결론적으로 어떻게 구현을 도와드릴수 있을지 궁금해서.. 이렇게 이슈로 남깁니다

정말 감이 안잡히는데 어떻게 해야 코어부분 개발을 도와드릴 수 있을까요?

**Comment by dlunch** [comment]
안녕하세요!!

프로젝트에 관심 가져주셔서 감사합니다. 

아직 시작 단계인데다가, 지금까지 1인 프로젝트에 가까웠어서 문서나 가이드같은 것이 없어 막막하게 느껴지실 수 있을 것 같네요.

일단 빠진 api구현부터 시작해 보시는 것이 어떨까요? 아직 구현되지 않은 api가 많아 기여하실 수 있는 부분이 많습니다.

적당한 앱을 하나 타겟으로 정하고(현재는 skt/ktf 앱만 제대로 로드가 됩니다), 실행해보면서 api가 없거나 구현이 잘못되어 있어 발생하는 에러를 보고 구현한다거나 하는 식으로 진행하시는 편이 작업하시는데 수월할 것 같습니다.

작업하시면서 질문 있으면 편하게 주세요.

감사합니다!

**Comment by mirusu400** [comment]
프로젝트의 어 부분이 API를 담당하는 부분일까요?

RustJava 부분을 건드려야되는건지, 아니면 wipi 부분 코드가 여기에 있는건지

그리고 어떻게 해당 부분을 구현하는지 (ex. 자바 애플리케이션을 리버싱해 실제 연산을 확인하고 구현해야하는지 등) 구체적으로 궁금합니다

감사합니다!

**Comment by dlunch** [comment, marked as answer]
구현 스펙에 대해서는 https://nikita36078.github.io/J2ME_Docs/docs/WIPI_API_1_1_1, http://strauss.cnu.ac.kr/research/wipi/download/WIPI%20V1.2.1_final(ST1.2.1).pdf, https://emulation.gametechwiki.com/index.php/Cellphone_emulators#Emulators_5 에서 받을 수 있는 emulator 등을 주로 참고하고 있습니다.

RustJava 쪽에는 WIPI가 기반하고 있는 java 1.2 api를 구현하고 있구요, wipi c/java 구현체가 이 리포지터리에 있습니다. wie_wipi_c, wie_wipi_java 등의 crate를 살펴보시면 좋을 것 같습니다.
